### DISCUSSION TITLE
Inquiry about `srsLagrange` in `plonk.setup`

### DISCUSSION THREAD

**Discussion by ZyadSamy** [Consensys/gnark#1134] [Q&A] [answered]
`plonk.setup` [function](https://github.com/Consensys/gnark/blob/master/backend/plonk/bn254/setup.go#L89) expects  `srs` and `srsLagrange`, how can i generate `srsLagrange` for a given SRS?

I tried the following

```
func ToLagrange(srs *kzg_bn254.SRS) (*kzg_bn254.SRS, error) {
	newSRS := &kzg_bn254.SRS{Vk: srs.Vk}

	g1, err := kzg_bn254.ToLagrangeG1(srs.Pk.G1)
	if err != nil {
		return nil, err
	}

	newSRS.Pk.G1 = g1

	return newSRS, nil
}
```

the setup function panics with an error like this:
`panic: kzg srs lagrange is too small: got 65536, need 32768`

**Comment by ThomasPiellard** [comment, marked as answer]
Hi, so clearly the message is misleading given the sizes, we need to change that. To convert an SRS to Lagrange you called the correct function, but you need to know the size of the constraint system, so you need to call `g1, err := kzg_bn254.ToLagrangeG1(srs.Pk.G1[:n])` where `n` is the size of the constraint system. The reason is that to convert an SRS to Lagrange form we use a FFT-like transformation where the exact size of the circuit is needed

**Reply by ivokub** [reply]
Also have a look at an [unit test](https://github.com/Consensys/gnark/blob/master/std/recursion/plonk/verifier_test.go#L360-L375) we use in PLONK recursion test for generating different PLONK verification keys using same KZG SRS.
